Mathlib/Combinatorics/SimpleGraph/Diameter.lean

```diff
@@ -0,0 +1,218 @@
+  ⨆ u, ⨆ v, G.edist u v
+
+lemma ediam_def : G.ediam = ⨆ p : α × α, G.edist p.1 p.2 := by
+  rw [ediam, iSup_prod]
+
+lemma edist_le_ediam {u v : α} : G.edist u v ≤ G.ediam :=
+  le_iSup₂ (f := G.edist) u v
+
+lemma ediam_eq_top_iff : G.ediam = ⊤ ↔ ∀ b < ⊤, ∃ u v, b < G.edist u v := by
+  simp only [ediam, iSup_eq_top, lt_iSup_iff]
+
+lemma ediam_eq_zero_of_subsingleton [Subsingleton α] : G.ediam = 0 := by
+  rw [ediam_def, ENat.iSup_eq_zero]
+  simpa [edist_eq_zero_iff, Prod.forall] using subsingleton_iff.mp ‹_›
+
+lemma nontrivial_of_ediam_ne_zero (h : G.ediam ≠ 0) : Nontrivial α := by
+  contrapose! h
+  rw [not_nontrivial_iff_subsingleton] at h
+  exact ediam_eq_zero_of_subsingleton
+
+lemma ediam_ne_zero [Nontrivial α] : G.ediam ≠ 0 := by
+  obtain ⟨u, v, huv⟩ := exists_pair_ne ‹_›
+  contrapose! huv
+  simp only [ediam, nonpos_iff_eq_zero, ENat.iSup_eq_zero, edist_eq_zero_iff] at huv
+  exact huv u v
+
+lemma subsingleton_of_ediam_eq_zero (h : G.ediam = 0) : Subsingleton α := by
+  contrapose! h
+  apply not_subsingleton_iff_nontrivial.mp at h
+  exact ediam_ne_zero
+
+lemma ediam_ne_zero_iff_nontrivial :
+    G.ediam ≠ 0 ↔ Nontrivial α :=
+  ⟨nontrivial_of_ediam_ne_zero, fun _ ↦ ediam_ne_zero⟩
+
+@[simp]
+lemma ediam_eq_zero_iff_subsingleton :
+    G.ediam = 0 ↔ Subsingleton α :=
+  ⟨subsingleton_of_ediam_eq_zero, fun _ ↦ ediam_eq_zero_of_subsingleton⟩
+
+lemma ediam_eq_top_of_not_connected [Nonempty α] (h : ¬G.Connected) : G.ediam = ⊤ := by
+  rw [connected_iff_exists_forall_reachable] at h
+  push_neg at h
+  obtain ⟨_, hw⟩ := h Classical.ofNonempty
+  rw [eq_top_iff, ← edist_eq_top_of_not_reachable hw]
+  exact edist_le_ediam
+
+lemma ediam_eq_top_of_not_preconnected (h : ¬G.Preconnected) : G.ediam = ⊤ := by
+  cases isEmpty_or_nonempty α
+  · exfalso
+    exact h <| IsEmpty.forall_iff.mpr trivial
+  · apply ediam_eq_top_of_not_connected
+    rw [connected_iff]
+    tauto
+
+lemma exists_edist_eq_ediam_of_ne_top [Nonempty α] (h : G.ediam ≠ ⊤) :
+    ∃ u v, G.edist u v = G.ediam :=
+  Prod.exists'.mp
+    <| ediam_def ▸ (ENat.sSup_mem_of_Nonempty_of_lt_top <| lt_top_iff_ne_top.mpr <| ediam_def ▸ h)
+
+lemma exists_edist_eq_ediam_of_finite [Nonempty α] [Finite α] :
+    ∃ u v, G.edist u v = G.ediam := by
+  let f : (α × α) → ℕ∞ := fun p ↦ G.edist p.1 p.2
+  by_cases h : G.ediam = ⊤
+  · have : ⨆ p, f p = ⊤ := by simp [f, ediam_def ▸ h]
+    have : ⊤ ∈ Set.range f := this ▸ (Set.range_nonempty _).csSup_mem (Set.finite_range _)
+    simp_all
+  · exact exists_edist_eq_ediam_of_ne_top h
+
+lemma ediam_mono_of_ne_top [Nonempty α] (h : G ≤ G') (hn : G'.ediam ≠ ⊤) :
+    G'.ediam ≤ G.ediam :=
+  have ⟨_, _, huv⟩ := G'.exists_edist_eq_ediam_of_ne_top hn
+  huv ▸ (edist_le_subgraph_edist h).trans edist_le_ediam
+
+lemma ediam_mono_of_finite [Nonempty α] [Finite α] (h : G ≤ G') :
+    G'.ediam ≤ G.ediam :=
+  have ⟨_, _, huv⟩ := G'.exists_edist_eq_ediam_of_finite
+  huv ▸ (edist_le_subgraph_edist h).trans edist_le_ediam
```

Comment:
You should stop thinking about the supremum as "either the maximum value of my function or infinity if my function is unbounded" and start thinking in terms of its universal property (which is usually all you need, really):
```suggestion
@[gcongr] lemma ediam_anti (h : G ≤ G') : G'.ediam ≤ G.ediam :=
  iSup₂_mono fun _ _ ↦ edist_le_subgraph_edist h
```
(btw `edist_le_subgraph_edist` should be called `edist_mono` and tagged `gcongr`, if you mind opening a PR!)

Reply:
I'm surprised I didn't notice that the hypotheses aren't needed. Thank you for the tip! I will keep it in mind. It highlights how `noncomputable` definitions can be more powerful than computational ones (to me, the universal property feels `noncomputable`-like, as opposed to the former which feels more computational).

Reply:
Yes, quite precisely!